subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/jar/SplitJarSnapshotCache.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.jar;
+
+import com.google.common.collect.Maps;
+import org.gradle.api.internal.changedetection.state.WellKnownFileLocations;
+import org.gradle.internal.Factory;
+import org.gradle.internal.hash.HashCode;
+
+import java.io.File;
+import java.util.Map;
+
+public class SplitJarSnapshotCache implements JarSnapshotCache {
+    private final WellKnownFileLocations fileLocations;
+    private final JarSnapshotCache globalCache;
+    private final JarSnapshotCache localCache;
+
+    public SplitJarSnapshotCache(WellKnownFileLocations fileLocations, JarSnapshotCache globalCache, JarSnapshotCache localCache) {
+        this.fileLocations = fileLocations;
+        this.globalCache = globalCache;
+        this.localCache = localCache;
+    }
+
+    @Override
+    public Map<File, JarSnapshot> getJarSnapshots(Map<File, HashCode> jars) {
+        Map<File, HashCode> globalJars = Maps.newLinkedHashMap();
+        Map<File, HashCode> localJars = Maps.newLinkedHashMap();
+        for (Map.Entry<File, HashCode> entry : jars.entrySet()) {
+            if (fileLocations.isImmutable(entry.getKey().getPath())) {
+                globalJars.put(entry.getKey(), entry.getValue());
+            } else {
+                localJars.put(entry.getKey(), entry.getValue());
+            }
+        }
+        Map<File, JarSnapshot> globalSnapshots = globalCache.getJarSnapshots(globalJars);
+        Map<File, JarSnapshot> localSnapshots = localCache.getJarSnapshots(localJars);
+
+        Map<File, JarSnapshot> snapshots = Maps.newLinkedHashMap();
+        for (File jar : jars.keySet()) {
+            JarSnapshot snapshot = globalSnapshots.get(jar);
+            if (snapshot == null) {
+                snapshot = localSnapshots.get(jar);
+            }
+            snapshots.put(jar, snapshot);
+        }
+        return snapshots;
```

Comment:
Is the iteration order of the resulting `Map` important? Otherwise, this loop is the same as

```java
Map<File, JarSnapshot> snapshots = localSnapshots;
snapshots.putAll(globalSnapshots);
return snapshots;
```

or to be on the safe side and not modify `localSnapshots`:

```java
return ImmutableMap.<File, JarSnapshot>builder()
            .putAll(localSnapshots)
            .putAll(globalSnapshots)
            .build();
```

Reply:
Good point, I'll check that

Reply:
It does matter, as we use it to determine if the classpath order has changed.

Reply:
Alright then, thanks for checking!